home/management/commands/updategeneds.py

```diff
@@ -0,0 +1,88 @@
+    help = '''Updates the database with new courses and professors during the provided semester.
+    The semester argument must be in the numerical form YEAR+SEASON (see ** for exception).
+    The season codes are as follows:
+        Spring -> 01
+        Summer -> 05
+         Fall  -> 08
+        Winter -> 12
+    EXAMPLE: Spring 2023 = 202301
+
+    ** NOTE: Starting from Winter 2021, the values for winter semesters are off by one year. Winter 2021 is actually 202012, not 202112
+   '''
+
+    def __init__(self, *args, **kwargs):
+        super().__init__()
+        self.num_geneds_updated = 0
+
+    def create_parser(self, *args, **kwargs):
+        # to format the help text that gets displayed with the -h option
+        parser = super(Command, self).create_parser(*args, **kwargs)
+        parser.formatter_class = RawTextHelpFormatter
+        return parser
+
+    def handle(self, *args, **options):
+        start = datetime.now()
+        pt_courses = Course.unfiltered.all().order_by('name')
+
+        # for each of our courses...
+        for pt_course in pt_courses:
+            # see if umdio has this course
+            umdio_course = requests.get(f"https://api.umd.io/v1/courses/{pt_course.name}").json()
+
+            # if umdio does not have this course continue to the next course
+            if isinstance(umdio_course, dict) and 'error_code' in umdio_course.keys():
+                continue
+
+            print(pt_course.name)
+            self.update_course_table(pt_course, umdio_course[0])
+            self.update_geneds_table(pt_course, umdio_course[0])
+
+        runtime = datetime.now() - start
+        print(f"number of geneds updated: {self.num_geneds_updated}")
+        print(f"total runtime: {round(runtime.seconds / 60, 2)} minutes")
+
+    def update_course_table(self, pt_course, umdio_course):
+        # umdio uses an empty list if a course has no geneds but
+        # we want to use null because that's what our schema is expecting.
+        if not umdio_course['gen_ed']:
+            pt_course.geneds = None
+        else:
+            pt_course.geneds = umdio_course['gen_ed']
+
+        pt_course.save()
+
+    def update_geneds_table(self, pt_course, umdio_course):
+        pt_geneds = Gened.objects.all()
+
+        # create a list of all the umdio geneds for this course.
+        # This list is considered the "correct" geneds for this course.
+        umdio_geneds = []
+        for gened_lst in umdio_course['gen_ed']:
+            for gened in gened_lst:
+                # To account for cases that have a pipe |, take the first 4
+                # characters of the gened value because geneds only have 4
+                # characters in their name
+                umdio_geneds.append(gened[:4])
+
+        # if we have a gened linked to a course but umdio doesn't agree,
+        # assume our records are outdated and delete this link.
+        for gened in pt_geneds.filter(course=pt_course):
```

Comment:
wouldn't `pt_course.gened_set` be better here? Or adjusting the foreign key related_name so we can use `pt_course.geneds`.

Reply:
`pt_course.geneds` is a json string which I thought would be more difficult to use than just getting the relevant geneds from the geneds table. I'm not following you with the `pt_course.gened_set`... what do you mean?

Reply:
`pt_course.gened_set` is a [`RelatedManager`](https://docs.djangoproject.com/en/4.1/ref/models/relations/), roughly equivalent to `pt_geneds.filter(course=pt_course)`. See eg: https://github.com/planetterp/PlanetTerp/blob/2e66d70a5226a69a8fc98f818725085cc61413d6/home/views/tools.py#L71

Reply:
I don't understand - you don't need to write any new code to enable this. Django automatically creates related managers for every foreign key. Swapping for `pt_course.gened_set.all()` will work as-is.

Reply:
I was just confused on the concept of related managers